src/cmap/conn.rs

```diff
@@ -327,16 +327,26 @@ impl PinnedConnectionHandle {
         }
     }
 
-    /// Retrieve the pinned connection, blocking until it's available for use.  Will fail if the
-    /// connection has been unpinned.
+    /// Retrieve the pinned connection  Will fail if the connection has been unpinned or is still in
+    /// use.
     pub(crate) async fn take_connection(&self) -> Result<PooledConnection> {
+        use tokio::sync::mpsc::error::TryRecvError;
         let mut receiver = self.receiver.lock().await;
-        let mut connection = receiver.recv().await.ok_or_else(|| {
-            Error::internal(format!(
-                "cannot take connection after unpin (id={})",
-                self.id
-            ))
-        })?;
+        let mut connection = match receiver.try_recv() {
```

Comment:
Using `try_recv` rather than `recv` means the re-entrant case becomes an error rather than a deadlock, which seems much better behavior to me.  The hypothetical downside is that it could fail in a race condition, but since any individual pinned connection is used in a strictly linear sequence that doesn't apply here.